### PR TITLE
Missing profile part

### PR DESCRIPTION
**Pull request by wbloszyk**
Reference to #614 #547 

## Subject
I add SonataProfileBundle - copy of missing part when update SonataUserBundle 3.x to 4.x. 

I am targeting this branch, because it is feature with BC. 

## Changelog

```markdown

### Added
- Added missing profile part when using SonataUserBundle > 3.x

### CODE REVIEW

File: src/ProfileBundle/Block/AccountBlockService.php [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+use Symfony\Component\OptionsResolver\OptionsResolver;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
+use Symfony\Component\Security\Core\User\UserInterface;
+
+/**
+ * Render a block with the connection option or the login name.
+ *
+ * @author Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ */
+class AccountBlockService extends AbstractAdminBlockService
```
Comment (core23) [inline src/ProfileBundle/Block/AccountBlockService.php]: Please make all classes final, if possible

File: src/ProfileBundle/Block/AccountBlockService.php [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+     */
+    public function __construct($name, EngineInterface $templating, TokenStorageInterface $tokenStorage)
+    {
+        parent::__construct($name, $templating);
+
+        $this->tokenStorage = $tokenStorage;
+    }
+
+    /**
+     * {@inheritdoc}
```
Comment (core23) [inline src/ProfileBundle/Block/AccountBlockService.php]: We don't need this

File: src/ProfileBundle/Block/AccountBlockService.php [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+{
+    /**
+     * @var TokenStorageInterface
+     */
+    private $tokenStorage;
+
+    /**
+     * @param string $name
+     */
+    public function __construct($name, EngineInterface $templating, TokenStorageInterface $tokenStorage)
```
Comment (core23) [inline src/ProfileBundle/Block/AccountBlockService.php]: Please use type hints

File: src/ProfileBundle/Controller/ProfileController.php [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+ *   - separation of the user authentication form with the profile form.
+ */
+class ProfileController extends Controller
+{
+    /**
+     * @throws AccessDeniedException
+     *
+     * @return Response
+     */
+    public function dashboardAction()
```
Comment (core23) [inline src/ProfileBundle/Controller/ProfileController.php]:
```suggestion
    public function dashboardAction(): Response
```

File: src/ProfileBundle/DependencyInjection/Configuration.php [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    {
+        $treeBuilder = new TreeBuilder();
+        $node = $treeBuilder->root('sonata_profile');
+
+        $node
+            ->children()
+                ->booleanNode('override_sonata_user_bundle_profile')->defaultValue(false)->end()
+                ->scalarNode('template')->defaultValue('SonataProfileBundle::action.html.twig')->end()
+                // Original code from the Sonata User Bundle 3.x
+                ->arrayNode('dashboard')
```
Comment (core23) [inline src/ProfileBundle/DependencyInjection/Configuration.php]: Please create small methods for each major node

File: src/ProfileBundle/Block/AccountBlockService.php [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+        return $this->renderPrivateResponse($blockContext->getTemplate(), [
+            'user' => $user,
+            'block' => $blockContext->getBlock(),
+            'context' => $blockContext,
+        ]);
+    }
+
+    /**
+     * {@inheritdoc}
```
Comment (core23) [inline src/ProfileBundle/Block/AccountBlockService.php]: Please remove all `inheritdoc`s

Comment (wbloszyk) [reply src/ProfileBundle/Block/AccountBlockService.php]: @core23 i remove all inheritdoc from my PR

File: src/ProfileBundle/Block/ProfileMenuBlockService.php [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+use Sonata\BlockBundle\Block\BlockContextInterface;
+use Sonata\BlockBundle\Block\Service\MenuBlockService;
+use Sonata\ProfileBundle\Menu\ProfileMenuBuilder;
+use Symfony\Bundle\FrameworkBundle\Templating\EngineInterface;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+/**
+ * @author Hugo Briand <briand@ekino.com>
+ */
+class ProfileMenuBlockService extends MenuBlockService
```
Comment (core23) [inline src/ProfileBundle/Block/ProfileMenuBlockService.php]:
```suggestion
final class ProfileMenuBlockService extends MenuBlockService
```

File: src/ProfileBundle/Block/ProfileMenuBlockService.php [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+{
+    /**
+     * @var ProfileMenuBuilder
+     */
+    private $menuBuilder;
+
+    /**
+     * ProfileMenuBlockService constructor.
```
Comment (core23) [inline src/ProfileBundle/Block/ProfileMenuBlockService.php]: No superflouse comments please

File: src/ProfileBundle/Controller/ProfileController.php [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Security\Core\Exception\AccessDeniedException;
+use Symfony\Component\Security\Core\User\UserInterface;
+
+/**
+ * This class is inspired from the FOS Profile Controller, except :
+ *   - only twig is supported
+ *   - separation of the user authentication form with the profile form.
+ */
+class ProfileController extends Controller
```
Comment (core23) [inline src/ProfileBundle/Controller/ProfileController.php]: Please make everything new final by default or use private properties where possible

File: src/ProfileBundle/Controller/ProfileController.php [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+{
+    /**
+     * @throws AccessDeniedException\
```
Comment (core23) [inline src/ProfileBundle/Controller/ProfileController.php]:
```suggestion
     * @throws AccessDeniedException
```

File: src/ProfileBundle/Controller/ProfileController.php (hunk 9 above)
Comment (wbloszyk) [reply src/ProfileBundle/Controller/ProfileController.php]: It is not new. I copy it from SonataUserBundle 3.x If someone override it when use SonataUserBundle 3.x, then can simple override it by change UserBundle to ProfileBundle

**Comment by wbloszyk** [comment]
@core23 @greg0ire @salamichel 

For now ProfileBundle is copy of missing profile from SonataUserBunde < 4.0. Base on [#1121](https://github.com/sonata-project/SonataUserBundle/pull/1121#issuecomment-568169326) SonataUserBundle should only integrate User to Admin, without frontend. Also ProfileBundle is not required in ecommerce bundle(but nice to have in sandbox). 

This bundle have:
- simple user block
- menu block
- profile component

Lets make new secound tier bundle with predefined block, components, tools etc., which is "Nice to have" in sandbox and other system, but is not required. By default should be disabled.

List what it can have:
- simple user block,
- menus tools with menu block (to define many menus)
- breadcrumb tools (simmilar to menu, also can be generated from menu)
- profiles components
- newsletter (block to display, controller to handler and events where developer can add what to do)
- newsletter in admin (to send email to subscibers)
- and many others

Mybe SonataToolsBundle? Or I should close this PR and replace this code to sandbox?

File: src/ProfileBundle/Block/AccountBlockService.php [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+ * @author Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ */
+final class AccountBlockService extends AbstractAdminBlockService
+{
+    /**
+     * @var TokenStorageInterface
+     */
+    private $tokenStorage;
+
+    public function __construct(string $name, EngineInterface $templating, TokenStorageInterface $tokenStorage)
```
Comment (core23) [inline src/ProfileBundle/Block/AccountBlockService.php]: A string name was first parameter is deprecated

File: src/ProfileBundle/DependencyInjection/Compiler/GlobalVariablesCompilerPass.php [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+namespace Sonata\ProfileBundle\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * @author Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ */
+class GlobalVariablesCompilerPass implements CompilerPassInterface
```
Comment (core23) [inline src/ProfileBundle/DependencyInjection/Compiler/GlobalVariablesCompilerPass.php]:
```suggestion
final class GlobalVariablesCompilerPass implements CompilerPassInterface
```

File: src/ProfileBundle/DependencyInjection/Configuration.php [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+use Symfony\Component\Config\Definition\Builder\ArrayNodeDefinition;
+use Symfony\Component\Config\Definition\Builder\TreeBuilder;
+use Symfony\Component\Config\Definition\ConfigurationInterface;
+
+/**
+ * This is the class that validates and merges configuration from your app/config files.
+ *
+ * To learn more see {@link http://symfony.com/doc/current/cookbook/bundles/extension.html#cookbook-bundles-extension-config-class}
+ */
+class Configuration implements ConfigurationInterface
```
Comment (core23) [inline src/ProfileBundle/DependencyInjection/Configuration.php]:
```suggestion
final class Configuration implements ConfigurationInterface
```

**Comment by wbloszyk** [comment]
@core23 
before ecommerce used profile from UserBundle. Now they should have own profile

**Comment by wbloszyk** [comment]
> So the whole profile stuff is not related to the ecommerce bundle at all? We'll create a new repo for this the next days

@core23 new repo or bundle for one feature is not enought. I resolve profile(dashboard) issue by add **template** and **menu_builder** variable to configuration. Now ecommerce have own profile, like UserBundle should have. Also profiles can be easy merge by change **menu_builder** options.

File: src/ProfileBundle/Block/AccountBlockService.php (hunk 11 above)
Comment (wbloszyk) [reply src/ProfileBundle/Block/AccountBlockService.php]: Ecommerce require:
        "sonata-project/block-bundle": "^3.8",
so a string name was required and should be deleted with BlockBundle dependencies update.

**Pull request closed without merging**